File: apps/desktop/src/lib/indicators.ts
/**
 * Technical Analysis - Types, Constants & Utilities
 *
 * All calculation functions have been migrated to Rust (src-tauri/src/indicators/).
 * Use indicators-rust.ts for async Tauri invoke wrappers.
 * This file retains only types, configuration constants, and convertToOHLC().
 */

// ============================================================================
// Data Types
// ============================================================================

export interface OHLCData {
  time: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume?: number;
}

export interface LineData {
  time: string;
  value: number | null;
}

export interface HistogramData {
  time: string;
  value: number | null;
  color?: string;
}

// ============================================================================
// Indicator Result Types
// ============================================================================

export interface MACDResult {
  macd: LineData[];
  signal: LineData[];
  histogram: HistogramData[];
}

export interface BollingerResult {
  upper: LineData[];
  middle: LineData[];
  lower: LineData[];
}

export interface StochasticResult {
  k: LineData[];  // %K (fast line)
  d: LineData[];  // %D (signal line)
}

export interface ADXResult {
  adx: LineData[];      // Trend strength (0-100)
  diPlus: LineData[];   // +DI
  diMinus: LineData[];  // -DI
}

export interface IchimokuResult {
  tenkan: LineData[];   // Conversion Line (9)
  kijun: LineData[];    // Base Line (26)
  senkouA: LineData[];  // Leading Span A (26 shifted forward)
  senkouB: LineData[];  // Leading Span B (52 shifted forward)
  chikou: LineData[];   // Lagging Span (26 shifted back)
}

export interface PivotPointsResult {
  pivot: LineData[];
  r1: LineData[];
  r2: LineData[];
  r3: LineData[];
  s1: LineData[];
  s2: LineData[];
  s3: LineData[];
}

export interface FibonacciLevel {
  level: number;
  price: number;
  label: string;
}

export interface FibonacciResult {
  levels: FibonacciLevel[];
  swingHigh: { price: number; time: string };
  swingLow: { price: number; time: string };
}

export interface AllIndicatorsResult {
  sma?: LineData[];
  ema?: LineData[];
  rsi?: LineData[];
  macd?: MACDResult;
  bollinger?: BollingerResult;
  atr?: LineData[];
  vwap?: LineData[];
  stochastic?: StochasticResult;
  obv?: LineData[];
  adx?: ADXResult;
  ichimoku?: IchimokuResult;
  pivot?: PivotPointsResult;
  fibonacci?: FibonacciResult;
  heikinAshi?: OHLCData[];
}

// ============================================================================
// Indicator Configuration
// ============================================================================

export type IndicatorType =
  | 'sma' | 'ema' | 'rsi' | 'macd' | 'bollinger' | 'atr' | 'vwap'
  | 'stochastic' | 'obv' | 'adx' | 'ichimoku' | 'pivot' | 'fibonacci';

export interface IndicatorConfig {
  id: string;
  type: IndicatorType;
  enabled: boolean;
  params: Record<string, number>;
  color?: string;
  /** For pivot points: 'standard' | 'fibonacci' | 'woodie' */
  pivotType?: 'standard' | 'fibonacci' | 'woodie';
}

export const defaultIndicatorConfigs: Record<IndicatorType, Omit<IndicatorConfig, 'id'>> = {
  sma: { type: 'sma', enabled: false, params: { period: 20 }, color: '#2196f3' },
  ema: { type: 'ema', enabled: false, params: { period: 20 }, color: '#ff9800' },
  rsi: { type: 'rsi', enabled: false, params: { period: 14 } },
  macd: { type: 'macd', enabled: false, params: { fast: 12, slow: 26, signal: 9 } },
  bollinger: { type: 'bollinger', enabled: false, params: { period: 20, stdDev: 2 }, color: '#9c27b0' },
  atr: { type: 'atr', enabled: false, params: { period: 14 } },
  vwap: { type: 'vwap', enabled: false, params: {}, color: '#e91e63' },
  stochastic: { type: 'stochastic', enabled: false, params: { kPeriod: 14, kSlowPeriod: 3, dPeriod: 3 } },
  obv: { type: 'obv', enabled: false, params: {} },
  adx: { type: 'adx', enabled: false, params: { period: 14 } },
  ichimoku: { type: 'ichimoku', enabled: false, params: { tenkan: 9, kijun: 26, senkouB: 52 }, color: '#00bcd4' },
  pivot: { type: 'pivot', enabled: false, params: {}, pivotType: 'standard' },
  fibonacci: { type: 'fibonacci', enabled: false, params: { lookback: 50 } },
};

// ============================================================================
// Signal Types (migrated from signals.ts)
// ============================================================================

export type SignalType =
  | 'rsi_oversold'
  | 'rsi_overbought'
  | 'macd_bullish_cross'
  | 'macd_bearish_cross'
  | 'bollinger_squeeze'
  | 'bollinger_breakout_up'
  | 'bollinger_breakout_down'
  | 'stochastic_oversold'
  | 'stochastic_overbought'
  | 'stochastic_bullish_cross'
  | 'stochastic_bearish_cross'
  | 'adx_trend_start'
  | 'adx_trend_strong'
  | 'golden_cross'
  | 'death_cross'
  | 'divergence_bullish'
  | 'divergence_bearish';

export type SignalDirection = 'bullish' | 'bearish' | 'neutral';
export type SignalStrength = 'strong' | 'moderate' | 'weak';

export interface TechnicalSignal {
  type: SignalType;
  direction: SignalDirection;
  strength: SignalStrength;
  date: string;
  price: number;
  indicator: string;
  value?: number;
  description: string;
}

export interface DivergenceSignal {
  type: 'bullish' | 'bearish';
  indicator: 'rsi' | 'macd' | 'obv' | 'stochastic';
  startDate: string;
  endDate: string;
  priceStart: number;
  priceEnd: number;
  indicatorStart: number;
  indicatorEnd: number;
  confidence: number;
}

export interface SignalDetectionConfig {
  rsiOversold?: number;
  rsiOverbought?: number;
  adxTrendThreshold?: number;
  adxStrongThreshold?: number;
  bollingerSqueezePeriod?: number;
  divergenceLookback?: number;
}

// ============================================================================
// Pattern Types (migrated from patterns.ts)
// ============================================================================

export type CandlestickPattern =
  | 'doji' | 'hammer' | 'inverted_hammer' | 'hanging_man'
  | 'shooting_star' | 'spinning_top' | 'marubozu_bullish' | 'marubozu_bearish'
  | 'engulfing_bullish' | 'engulfing_bearish' | 'harami_bullish' | 'harami_bearish'
  | 'piercing_line' | 'dark_cloud_cover' | 'tweezer_top' | 'tweezer_bottom'
  | 'morning_star' | 'evening_star' | 'three_white_soldiers' | 'three_black_crows'
  | 'three_inside_up' | 'three_inside_down';

export type PatternDirection = 'bullish' | 'bearish' | 'neutral';
export type PatternReliability = 'high' | 'medium' | 'low';

export interface PatternMatch {
  pattern: CandlestickPattern;
  name: string;
  startIndex: number;
  endIndex: number;
  direction: PatternDirection;
  reliability: PatternReliability;
  description: string;
}

// ============================================================================
// Trading Analysis Types (Regime, Setup Scoring, Risk)
// ============================================================================

export type RegimeType =
  | 'strong_uptrend' | 'uptrend' | 'sideways'
  | 'downtrend' | 'strong_downtrend' | 'squeeze';

export type VolatilityLevel = 'low' | 'normal' | 'high' | 'extreme';

export type SetupType =
  | 'trend_following' | 'breakout' | 'reversal'
  | 'mean_reversion' | 'squeeze' | 'no_setup';

export type TradeDirection = 'long' | 'short' | 'neutral';

export interface RegimeAnalysis {
  regime: RegimeType;
  confidence: number;
  trendStrength: number;
  volatilityLevel: VolatilityLevel;
  momentum: number;
  supportingEvidence: string[];
}

export interface SetupFactorDetail {
  name: string;
  score: number;
  weight: number;
  description: string;
}

export interface SetupScore {
  totalScore: number;
  regimeScore: number;
  momentumScore: number;
  patternScore: number;
  volumeScore: number;
  riskScore: number;
  factors: SetupFactorDetail[];
  setupType: SetupType;
  setupLabel: string;
  direction: TradeDirection;
}

export interface RiskAnalysis {
  atrValue: number;
  atrStopPrice: number;
  atrStopPercent: number;
  suggestedPositionSize: number;
  suggestedPositionValue: number;
  riskPerTrade: number;
  riskRewardRatio: number;
  portfolioImpactPercent: number;
  maxLossPercent: number;
  entryPrice: number;
  stopPrice: number;
  targetPrice: number;
}

export interface TradingAnalysis {
  regime: RegimeAnalysis;
  setup: SetupScore;
  risk: RiskAnalysis | null;
}

// ============================================================================
// Utility: Convert price-only data to synthetic OHLC
// ============================================================================

export function convertToOHLC(
  priceData: Array<{ date: string; value: number }>,
  volatilityPercent: number = 1.5
): OHLCData[] {
  return priceData.map((d, i) => {
    const close = d.value;
    const variance = close * (volatilityPercent / 100);
    const open = i > 0 ? priceData[i - 1].value : close;
    const high = Math.max(open, close) + Math.random() * variance;
    const low = Math.min(open, close) - Math.random() * variance;

    return {
      time: d.date,
      open,
      high,
      low,
      close,
      volume: Math.floor(Math.random() * 1000000) + 100000,
    };
  });
}

File: apps/desktop/src-tauri/src/commands/indicators.rs
use crate::indicators::types::*;
use crate::indicators::calculations;
use crate::indicators::patterns;
use crate::indicators::signals;
use crate::indicators::screener;

// ============================================================================
// Individual Indicator Commands
// ============================================================================

#[tauri::command]
pub fn calculate_sma(data: Vec<OHLCData>, period: usize) -> Vec<LineData> {
    calculations::calculate_sma(&data, period)
}

#[tauri::command]
pub fn calculate_ema(data: Vec<OHLCData>, period: usize) -> Vec<LineData> {
    calculations::calculate_ema(&data, period)
}

#[tauri::command]
pub fn calculate_rsi(data: Vec<OHLCData>, period: Option<usize>) -> Vec<LineData> {
    calculations::calculate_rsi(&data, period.unwrap_or(14))
}

#[tauri::command]
pub fn calculate_macd(data: Vec<OHLCData>, fast: Option<usize>, slow: Option<usize>, signal: Option<usize>) -> MACDResult {
    calculations::calculate_macd(&data, fast.unwrap_or(12), slow.unwrap_or(26), signal.unwrap_or(9))
}

#[tauri::command]
pub fn calculate_bollinger(data: Vec<OHLCData>, period: Option<usize>, std_dev: Option<f64>) -> BollingerResult {
    calculations::calculate_bollinger(&data, period.unwrap_or(20), std_dev.unwrap_or(2.0))
}

#[tauri::command]
pub fn calculate_atr(data: Vec<OHLCData>, period: Option<usize>) -> Vec<LineData> {
    calculations::calculate_atr(&data, period.unwrap_or(14))
}

#[tauri::command]
pub fn calculate_vwap(data: Vec<OHLCData>) -> Vec<LineData> {
    calculations::calculate_vwap(&data)
}

#[tauri::command]
pub fn calculate_stochastic(data: Vec<OHLCData>, k_period: Option<usize>, k_slow: Option<usize>, d_period: Option<usize>) -> StochasticResult {
    calculations::calculate_stochastic(&data, k_period.unwrap_or(14), k_slow.unwrap_or(3), d_period.unwrap_or(3))
}

#[tauri::command]
pub fn calculate_obv(data: Vec<OHLCData>) -> Vec<LineData> {
    calculations::calculate_obv(&data)
}

#[tauri::command]
pub fn calculate_adx(data: Vec<OHLCData>, period: Option<usize>) -> ADXResult {
    calculations::calculate_adx(&data, period.unwrap_or(14))
}

#[tauri::command]
pub fn calculate_ichimoku(data: Vec<OHLCData>, tenkan: Option<usize>, kijun: Option<usize>, senkou_b: Option<usize>) -> IchimokuResult {
    calculations::calculate_ichimoku(&data, tenkan.unwrap_or(9), kijun.unwrap_or(26), senkou_b.unwrap_or(52))
}

#[tauri::command]
pub fn calculate_pivot_points(data: Vec<OHLCData>, pivot_type: Option<String>) -> PivotPointsResult {
    calculations::calculate_pivot_points(&data, &pivot_type.unwrap_or_else(|| "standard".to_string()))
}

#[tauri::command]
pub fn calculate_fibonacci(data: Vec<OHLCData>, lookback: Option<usize>) -> FibonacciResult {
    calculations::calculate_fibonacci(&data, lookback.unwrap_or(50))
}

#[tauri::command]
pub fn convert_to_heikin_ashi(data: Vec<OHLCData>) -> Vec<OHLCData> {
    calculations::convert_to_heikin_ashi(&data)
}

// ============================================================================
// Batch: Calculate all requested indicators in one call
// ============================================================================

#[derive(serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct IndicatorRequest {
    pub indicator_type: String,
    pub params: std::collections::HashMap<String, serde_json::Value>,
}

#[tauri::command]
pub fn calculate_all_indicators(data: Vec<OHLCData>, requests: Vec<IndicatorRequest>) -> AllIndicatorsResult {
    let mut result = AllIndicatorsResult {
        sma: None, ema: None, rsi: None, macd: None, bollinger: None,
        atr: None, vwap: None, stochastic: None, obv: None, adx: None,
        ichimoku: None, pivot: None, fibonacci: None, heikin_ashi: None,
    };

    for req in &requests {
        match req.indicator_type.as_str() {
            "sma" => {
                let period = req.params.get("period").and_then(|v| v.as_u64()).unwrap_or(20) as usize;
                result.sma = Some(calculations::calculate_sma(&data, period));
            }
            "ema" => {
                let period = req.params.get("period").and_then(|v| v.as_u64()).unwrap_or(20) as usize;
                result.ema = Some(calculations::calculate_ema(&data, period));
            }
            "rsi" => {
                let period = req.params.get("period").and_then(|v| v.as_u64()).unwrap_or(14) as usize;
                result.rsi = Some(calculations::calculate_rsi(&data, period));
            }
            "macd" => {
                let fast = req.params.get("fast").and_then(|v| v.as_u64()).unwrap_or(12) as usize;
                let slow = req.params.get("slow").and_then(|v| v.as_u64()).unwrap_or(26) as usize;
                let signal = req.params.get("signal").and_then(|v| v.as_u64()).unwrap_or(9) as usize;
                result.macd = Some(calculations::calculate_macd(&data, fast, slow, signal));
            }
            "bollinger" => {
                let period = req.params.get("period").and_then(|v| v.as_u64()).unwrap_or(20) as usize;
                let std_dev = req.params.get("stdDev").and_then(|v| v.as_f64()).unwrap_or(2.0);
                result.bollinger = Some(calculations::calculate_bollinger(&data, period, std_dev));
            }
            "atr" => {
                let period = req.params.get("period").and_then(|v| v.as_u64()).unwrap_or(14) as usize;
                result.atr = Some(calculations::calculate_atr(&data, period));
            }
            "vwap" => {
                result.vwap = Some(calculations::calculate_vwap(&data));
            }
            "stochastic" => {
                let k = req.params.get("kPeriod").and_then(|v| v.as_u64()).unwrap_or(14) as usize;
                let ks = req.params.get("kSlowPeriod").and_then(|v| v.as_u64()).unwrap_or(3) as usize;
                let d = req.params.get("dPeriod").and_then(|v| v.as_u64()).unwrap_or(3) as usize;
                result.stochastic = Some(calculations::calculate_stochastic(&data, k, ks, d));
            }
            "obv" => {
                result.obv = Some(calculations::calculate_obv(&data));
            }
            "adx" => {
                let period = req.params.get("period").and_then(|v| v.as_u64()).unwrap_or(14) as usize;
                result.adx = Some(calculations::calculate_adx(&data, period));
            }
            "ichimoku" => {
                let tenkan = req.params.get("tenkan").and_then(|v| v.as_u64()).unwrap_or(9) as usize;
                let kijun = req.params.get("kijun").and_then(|v| v.as_u64()).unwrap_or(26) as usize;
                let senkou_b = req.params.get("senkouB").and_then(|v| v.as_u64()).unwrap_or(52) as usize;
                result.ichimoku = Some(calculations::calculate_ichimoku(&data, tenkan, kijun, senkou_b));
            }
            "pivot" => {
                let pt = req.params.get("pivotType").and_then(|v| v.as_str()).unwrap_or("standard");
                result.pivot = Some(calculations::calculate_pivot_points(&data, pt));
            }
            "fibonacci" => {
                let lb = req.params.get("lookback").and_then(|v| v.as_u64()).unwrap_or(50) as usize;
                result.fibonacci = Some(calculations::calculate_fibonacci(&data, lb));
            }
            "heikinAshi" => {
                result.heikin_ashi = Some(calculations::convert_to_heikin_ashi(&data));
            }
            _ => {}
        }
    }

    result
}

// ============================================================================
// Pattern Detection Command
// ============================================================================

#[tauri::command]
pub fn detect_candlestick_patterns(data: Vec<OHLCData>) -> Vec<PatternMatch> {
    patterns::detect_candlestick_patterns(&data)
}

// ============================================================================
// Signal Detection Commands
// ============================================================================

#[tauri::command]
pub fn detect_signals(data: Vec<OHLCData>, config: Option<SignalDetectionConfig>) -> Vec<TechnicalSignal> {
    let cfg = config.unwrap_or_default();
    signals::detect_signals(&data, &cfg)
}

#[tauri::command]
pub fn get_all_signals(data: Vec<OHLCData>, config: Option<SignalDetectionConfig>) -> Vec<TechnicalSignal> {
    let cfg = config.unwrap_or_default();
    signals::get_all_signals(&data, &cfg)
}

#[tauri::command]
pub fn detect_all_divergences(data: Vec<OHLCData>, lookback: Option<usize>) -> Vec<DivergenceSignal> {
    signals::detect_all_divergences(&data, lookback.unwrap_or(20))
}

// ============================================================================
// Screener Command
// ============================================================================

#[tauri::command]
pub fn run_screener(securities: Vec<ScreenerSecurityData>, filters: Vec<ScreenerFilter>) -> Vec<ScreenerResult> {
    screener::run_screener(&securities, &filters)
}
